pkg/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,89 @@
+package garbagecollector
+
+import (
+	"path"
+	"time"
+
+	"github.com/golang/glog"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	winformers "github.com/sdminonne/workflow-controller/pkg/client/informers/externalversions"
+	wlisters "github.com/sdminonne/workflow-controller/pkg/client/listers/workflow/v1"
+	wclientset "github.com/sdminonne/workflow-controller/pkg/client/versioned"
+
+	"github.com/sdminonne/workflow-controller/pkg/controller"
+)
+
+const (
+	// Interval represent the interval to run Garabge Collection
+	Interval time.Duration = 10 * time.Second
+)
+
+// GarbageCollector represents a Workflow Garbage Collector.
+// It collects orphaned Jobs
+type GarbageCollector struct {
+	KubeClient     clientset.Interface
+	WorkflowClient wclientset.Interface
+	WorkflowLister wlisters.WorkflowLister
+	WorkflowSynced cache.InformerSynced
+}
+
+// NewGarbageCollector builds initializes and returns a GarbageCollector
+func NewGarbageCollector(workflowClient wclientset.Interface, kubeClient clientset.Interface, workflowInformerFactory winformers.SharedInformerFactory) *GarbageCollector {
+	return &GarbageCollector{
+		KubeClient:     kubeClient,
+		WorkflowClient: workflowClient,
+		WorkflowLister: workflowInformerFactory.Workflow().V1().Workflows().Lister(),
+		WorkflowSynced: workflowInformerFactory.Workflow().V1().Workflows().Informer().HasSynced,
+	}
+}
+
+// CollectWorkflowJobs collect the orphaned jobs. First looking in the workflow informer list
+// then retrieve from the API and in case NotFound then remove via DeleteCollection primitive
+func (c *GarbageCollector) CollectWorkflowJobs() {
+	glog.V(4).Infof("Collecting garbage jobs")
+	jobs, err := c.KubeClient.BatchV1().Jobs(metav1.NamespaceAll).List(metav1.ListOptions{
+		LabelSelector: controller.WorkflowLabelKey,
+	})
+	if err != nil {
+		glog.Errorf("Garbage collector: unable to list workflow jobs to be collected: %v", err)
+	}
+	collected := make(map[string]struct{})
+	for _, job := range jobs.Items {
+		workflowName, found := job.Labels[controller.WorkflowLabelKey]
+		if !found {
```

Comment:
Does this test make sense? if it fails it means that the Jobs lister did return an items that does not respect the LabelSelector... any corner case that I am missing here?

Reply:
The day someone will break Generalized Label Selector for CRD it will break.